doc/mpc_node_specification.md

```diff
@@ -0,0 +1,370 @@
+   and would no longer have it when the delayed `INIT` message arrives. But
+   thanks to the `Prepare` state, we know all participants are at least in the
+   `Prepare` phase and thus have marked the input Triples as `Using` before the
+   owner can even start running the protocol. Hence, once it reaches `done`,
+   every participant has it as `Using`. In that state, peers never delete
+   triples since it knows the protocol is running and it should be expected that
+   the owner may finish faster than the peer.
+
+
+#### Running
+
+In the `Running` state, nodes continuously poke the local message generators and
+exchange messages to progress the protocol.
+
+Messages sent to participants that are offline or reject the message must be
+retried until either the receiver accepts the message or the protocol invocation
+globally times out.
+
+For recovery, we use the
+[Event-Sourcing](https://dev.to/alisamir/understanding-event-sourcing-a-detailed-guide-4cjp)
+pattern, while the protocol is running. The actual latest state is only held in
+memory but the stored events allow to recover it from persisted data.
+
+Specifically, this means tha all received messages during the `Running` state
+must be persisted before they are applied. These can be replayed on boot for
+recovery after a crash. After a protocol finishes, events can be discarded.
+
+Note: Unlike fully event-driven architectures, the MPC network does not need to
+have a separate even store component. The Event-Sourcing pattern is only applied
+locally within one node.
+
+### Owner Transitions
+
+The owner of the protocol invocation progresses the protocol:
+
+- When a new valid request is registered (indexed for the signature protocol,
+  initiated by background task for triple/pre-signature generation)
+    - Select a list of potential participants, based on the own view of which
+      nodes are active and hold the relevant data in their storage.
+    - Asynchronously send an INIT message to all the participants and wait until
+      enough messages have been received. (HTTP success response or explicit
+      OK message back)
+    - Transition to `Prepare` state
+- Once enough INIT messages have been answered with a success response
+    - Wait for finality of the request (for signatures only)
+    - Select participants that responded OK
+    - Send out a START message to all participants, with the list of participants
+    - Transition to `Running` state
+    - Start poking own generator and sending out the corresponding messages
+
+Owners must also handle the following error case:
+- Not enough participants answered the INIT message with an HTTP success response
+    - Retry, maybe some more participants are online now
+    - After a threshold of retries, switch to a different P or T input that has
+      more active participants.
+
+Other error cases can be handled with a global timeout to abort and retry for
+signatures.
```

Comment:
So additionally with https://github.com/sig-net/mpc/pull/352, other participants can send rejections in the case they don't agree for whatever reason like proposing an id that was already owned in the system by somebody else.

It might also be hard to do acknowledgements based on successful HTTP responses since the messaging layer currently handles that, which is why receiving an `Accept` would be better for now. At least its easier to implement that way

Reply:
I see. Updated the spec to make it either an HTTP response or a separate message.

The reason why an HTTP response would be preferred is to reduce latency and send overall less messages. HTTP already comes with a response, so I thought we could use that.

But yeah I can see that it's easier to implement with a separate `Accept` message. Let's do that for now and keep the HTTP response as an option in the future, when we have the luxury to micro-optimize :)